include/RAJA/policy/cuda/reduce.hpp

```diff
@@ -109,1507 +390,1185 @@ __device__ __forceinline__ T shfl_xor(T var, int laneMask)
+  //! transfers from the host to the device
+  RAJA_INLINE
+  void hostToDevice(Offload_Info &)
   {
-#if defined(__CUDA_ARCH__)
-    if (m_finish_reduction) {
-      extern __shared__ unsigned char sd_block[];
-      T *sd = reinterpret_cast<T *>(&sd_block[m_smem_offset]);
+  }
 
-      int threadId = threadIdx.x + blockDim.x * threadIdx.y
-                     + (blockDim.x * blockDim.y) * threadIdx.z;
+  //! transfers from the device to the host
+  RAJA_INLINE
+  void deviceToHost(Offload_Info &)
+  {
+    auto end = tally.list->streamEnd();
+    for(auto s = tally.list->streamBegin(); s != end; ++s) {
+      synchronize(*s);
+    }
+  }
 
-      __syncthreads();
+  //! frees all data used
+  //  frees all values in the pinned tally
+  RAJA_INLINE
+  void cleanup(Offload_Info &)
+  {
+    tally.list->free_list();
+  }
+};
 
-      for (int i = BLOCK_SIZE / 2; i >= WARP_SIZE; i /= 2) {
-        if (threadId < i) {
-          sd[threadId] = RAJA_MIN(sd[threadId], sd[threadId + i]);
-        }
-        __syncthreads();
-      }
 
-      for (int i = WARP_SIZE / 2; i > 0; i /= 2) {
-        if (threadId < i) {
-          sd[threadId] = RAJA_MIN(sd[threadId], sd[threadId + i]);
-        }
-      }
+//! Reduction data for Cuda Offload -- stores value, host pointer
+template <bool Async, typename Reducer, typename T>
+struct ReduceAtomic_Data {
+  //! union to hold either pointer to PinnedTally or poiter to value
+  //  only use list before setup for device and only use val_ptr after
+  union tally_u {
+    PinnedTally<T>* list;
+    T *val_ptr;
+    tally_u(PinnedTally<T>* l) : list(l) {};
+    tally_u(T *v_ptr) : val_ptr(v_ptr) {};
+  };
+  
+  mutable T value;
+  tally_u tally;
+  unsigned int* device_count;
+  T* device;
+  bool own_device_ptr;
+
+  //! disallow default constructor
+  ReduceAtomic_Data() = delete;
+
+  /*! \brief create from a default value and offload information
+   *
+   *  allocates PinnedTally to hold device values
+   */
+  explicit ReduceAtomic_Data(T initValue)
+      : value{initValue},
+        tally{new PinnedTally<T>},
+        device_count{nullptr},
+        device{nullptr},
+        own_device_ptr{false}
+  {
+  }
 
-      if (threadId < 1) {
-        _atomicMin<T>(&m_tally_device->tally, sd[threadId]);
-      }
+  RAJA_HOST_DEVICE
+  ReduceAtomic_Data(const ReduceAtomic_Data &other)
+      : value{Reducer::identity},
+        tally{other.tally},
+        device_count{other.device_count},
+        device{other.device},
+        own_device_ptr{false}
+  {
+  }
+  
+  //! delete pinned tally
+  RAJA_INLINE
+  void destroy()
+  {
+    delete tally.list; tally.list = nullptr;
+  }
+
+  //! check and setup for device
+  //  allocate device pointers and get a new result buffer from the pinned tally
+  RAJA_INLINE
+  bool setupForDevice(Offload_Info &info)
+  {
+    bool act = !device && setupReducers();
+    if (act) {
+      device = device_mempool_type::getInstance().malloc<T>(1);
```

Comment:
Why do we need a device_mempool_type and device_zeroed_mempool_type? Is the overhead really that bad? can't we just zero it out?